cpp/src/join/join_kernels.cuh

```diff
@@ -382,6 +424,125 @@ __global__ void probe_hash_table(multimap_type multi_map,
   }
 }
 
+/* --------------------------------------------------------------------------*/
+/**
+ * @brief  Performs a nested loop join to find all matching rows
+ between the left and right tables and generate the output for the desired Join operation.
+ *
+ * @param[in] left_table The left table
+ * @param[in] right_table The right table
+ * @param[in] check_row_equality The row equality comparator
+ * @param[in] left_num_rows The number of rows in the left table
+ * @param[in] right_num_rows The number of rows in the right table
+ * @param[out] join_output_l The left result of the join operation
+ * @param[out] join_output_r The right result of the join operation
+ * @param[in,out] current_idx A global counter used by threads to coordinate writes to the global
+ output
+ * @param[in] max_size The maximum size of the output
+ * @param[in] flip_results Whether to flip left and right outputs
+ * @param[in] offset An optional offset
+ * @tparam JoinKind The type of join to be performed
+ * @tparam block_size The number of threads per block for this kernel
+ * @tparam output_cache_size The side of the shared memory buffer to cache join output results
+ *
+ */
+/* ----------------------------------------------------------------------------*/
+template <join_kind JoinKind, cudf::size_type block_size, cudf::size_type output_cache_size>
```

Comment:
I'm not sure if `JoinKind `should be a template parameter. It increases the code size/compilation time 4x and the parameter is only used in a single condition. I would suggest making it a function parameter and checking if it causes a noticeable performance regressions.